main.cpp

```diff
@@ -1,55 +1,115 @@
+template<template<typename...> typename TypeSet, typename ...Ts1, typename ...Ts2, typename... Rest>
+constexpr auto concat(TypeSet<Ts1...>, TypeSet<Ts2...>, Rest...) {    //类模板参数拼接
+	return concat(TypeSet<Ts1..., Ts2...>{}, Rest{}...);
+}
+
+template<template<typename, typename...> typename TypeSet, typename T, typename ...Rest>  //类模板参数去重
+constexpr auto make_unique_typeset(TypeSet<T, Rest...>) {
+	if constexpr ((std::same_as<T, Rest> || ...)) {
+		return make_unique_typeset(TypeSet<Rest... >{});
+	}
+	else {
+		if constexpr (sizeof...(Rest) > 0) {
+			return concat(TypeSet<T>{}, make_unique_typeset(TypeSet<Rest...>{}));
+		}
+		else {
+			return TypeSet<T>{};
+		}
+	}
+}
+
+template <class T1, class T2> requires Addable<T1, T2>   //向量加法，要求向量元素是可加的
+auto operator+(std::vector<T1> const& a, std::vector<T2> const& b) {
+	using TR = std::decay_t<decltype(std::declval<T1>() + std::declval<T2>())>;
+	std::vector<TR> res;
+	const auto size = std::min(a.size(), b.size());
+	res.reserve(size);
+	for (size_t i = 0; i < size; i++) {
+		res.emplace_back(a[i] + b[i]);
+	}
+	return res;
+	// 请实现列表的逐元素加法！10 分
+	// 例如 {1, 2} + {3, 4} = {4, 6}
 }
 
-template <class T1, class T2>
-std::variant<T1, T2> operator+(std::variant<T1, T2> const &a, std::variant<T1, T2> const &b) {
-    // 请实现自动匹配容器中具体类型的加法！10 分
+template <class TR, class ...Ts> requires (Addable<Ts, TR> && ...) && (!Same<TR, std::variant<Ts...>>)
+auto operator+(std::variant<Ts...> const& a, TR const& b) {  //variant对象+对象，要求variant的每个子类型和对象都是可加的
+	std::variant<std::decay_t<decltype(std::declval<Ts>() + std::declval<TR>())>...> new_res;  //variant通过加法新生成的类型
+	auto res = make_unique_typeset(concat(std::variant<Ts...>{}, new_res));  //variant新旧类型进行拼接，然后去重
+	std::visit([&](auto&& arg) {res = arg + b; }, a);
+	return res;
+	// 请实现自动匹配容器中具体类型的加法！10 分
 }
 
-template <class T1, class T2>
-std::ostream &operator<<(std::ostream &os, std::variant<T1, T2> const &a) {
-    // 请实现自动匹配容器中具体类型的打印！10 分
+template <class TL, class ...Ts> requires (Addable<TL, Ts> && ...) && (!Same<TL, std::variant<Ts...>>)
+auto operator+(TL const& a, std::variant<Ts...> const& b) {  //对象+variant对象，要求对象和variant的每个子类型都是可加的
+	std::variant<std::decay_t<decltype(std::declval<TL>() + std::declval<Ts>())>...> new_res;  //variant通过加法新生成的类型
+	auto res = make_unique_typeset(concat(std::variant<Ts...>{}, new_res));  //variant新旧类型进行拼接，然后去重
+	std::visit([&](auto&& arg) {res = a + arg; }, b);
+	return res;
+	// 请实现自动匹配容器中具体类型的加法！10 分
 }
 
+template <class ...TLs, class ...TRs> requires (Addable<TLs, std::variant<TRs...>> && ...) && (Addable<std::variant<TLs...>, TRs> && ...)
+auto operator+(std::variant<TLs...> const& a, std::variant<TRs...> const& b) { //两个variant对象相加，要求第一个variant的任意一个子类型与第二个variant的任意一个子类型可加
+	auto res = make_unique_typeset(
+		concat(a, b, (TLs{} + std::variant<TRs...>{})..., (std::variant<TLs...>{} + TRs{})...));  //variant新旧类型进行拼接，然后去重
+	std::visit([&](auto&& arg1, auto&& arg2) {res = arg1 + arg2; }, a, b);
+	return res;
```

Comment:
```suggestion
	using Res = decltype(make_unique_typeset(concat(a, b, (std::declval<TLs>() + std::declval<std::variant<TRs...>>())..., (std::declval<std::variant<TLs...>>() + std::declval<TRs>())...)));  //variant新旧类型进行拼接，然后去重
	return std::visit([&](auto&& arg1, auto&& arg2) { return Res{arg1 + arg2}; }, a, b);
```
std::declval 从虚空中创建了一个对象，但是只能用于 decltype 中的非求值上下文中，从而不需要 TLs, TRs 具有默认构造函数。

Reply:
改成直接使用`std::visit`返回值的版本了，之前居然没有想到Orz